piff/knn_interp.py

```diff
@@ -0,0 +1,373 @@
+            # if no y, then interpolate
+            y = numpy.array([self.knn[key].predict(X) for key in self.attr_target]).T
+        if logger:
+            logger.debug('Regression shape: %s', y.shape)
+        # add misalignment shape (n_targets, 3)
+        # X is (n_samples, 2)
+        # y is (n_samples, n_targets)
+        y = y + self.misalignment[numpy.newaxis, :, 0] \
+                + X[:, 1, numpy.newaxis] * self.misalignment[numpy.newaxis, :, 1] \
+                + X[:, 0, numpy.newaxis] * self.misalignment[numpy.newaxis, :, 2]
+
+        return y
+
+    def writeSolution(self, fits, extname):
+        """Write the solution to a FITS binary table.
+
+        Save the knn params and the X and y arrays
+
+        :param fits:        An open fitsio.FITS object.
+        :param extname:     The name of the extension with the interp information.
+        """
+
+        dtypes = [('X', self.X.dtype, self.X.shape), ('Y', self.y.dtype, self.y.shape),
+                  ('XPIXEL', self.Xpixel.dtype, self.Xpixel.shape),
+                  ('ATTR_TARGET', self.attr_target.dtype, self.attr_target.shape),
+                  ('ATTR_INTERP', self.attr_interp.dtype, self.attr_interp.shape),
+                  ('MISALIGNMENT', self.misalignment.dtype, self.misalignment.shape)]
+        data = numpy.empty(1, dtype=dtypes)
+        # assign
+        data['X'] = self.X
+        data['Y'] = self.y
+        data['XPIXEL'] = self.Xpixel
+        data['ATTR_TARGET'] = self.attr_target
+        data['ATTR_INTERP'] = self.attr_interp
+        data['MISALIGNMENT'] = self.misalignment
+
+        # put the knn params in the header?
+        header = self.kwargs
+
+        # write to fits
+        fits.write_table(data, extname=extname, header=header)
+
+    def readSolution(self, fits, extname):
+        """Read the solution from a FITS binary table.
+
+        The extension should contain the same values as are saved
+        in the writeSolution method.
+
+        :param fits:        An open fitsio.FITS object.
+        :param extname:     The name of the extension with the interp information.
+        """
+        # header = fits[extname].read_header()
+        data = fits[extname].read()
+
+        # kwargs come from base class read()
+        # self.kwargs = header
+
+        # attr_target and attr_interp assigned in build
+        self.build(data['ATTR_INTERP'][0], data['ATTR_TARGET'][0])
+        # self.X and self.y assigned in _fit
+        self._fit(data['X'][0], data['Y'][0])
+        self.misalign_wavefront(data['MISALIGNMENT'][0])
+
+        # other attributes
+        self.Xpixel = data['XPIXEL'][0]
+
+
+# pretty sure the correct way to do this is in the WCS framework, but let's
+# move forward first:
+def pixel_to_focal(stardata):
+    """Take stardata and add focal plane position to properties
+
+    :param stardata:    The stardata with property 'ccdnum'
+
+    :returns stardata:  New stardata with updated properties
+    """
```

Comment:
I am pretty sure that the nominal pointing is not sufficient.  I’ve looked at the Blanco pointing, recently in fact, and it is only good to 10-15 arcsec, with some long tails too.




On Jul 5, 2016, at 7:14 PM, Mike Jarvis <notifications@github.com<mailto:notifications@github.com>> wrote:


In piff/knn_interp.py<https://github.com/rmjarvis/Piff/pull/24#discussion_r69611430>:

> +        self._fit(data['X'][0], data['Y'][0])
> +        self.misalign_wavefront(data['MISALIGNMENT'][0])
> +
> +        # other attributes
> +        self.Xpixel = data['XPIXEL'][0]
> +
> +
> +# pretty sure the correct way to do this is in the WCS framework, but let's
> +# move forward first:
> +def pixel_to_focal(stardata):
> +    """Take stardata and add focal plane position to properties
> +
> +    :param stardata:    The stardata with property 'ccdnum'
> +
> +    :returns stardata:  New stardata with updated properties
> +    """


The only question I guess is whether getting the pointing value where (u,v) == (0,0) is expected to be reliable enough for you using the nominal telescope pointing in the fits header.

If so, using the regular wcs and (u,v) is probably easier in the long run. But if we're worried about the origin of the coordinate system moving around too much because of approximations in the telescope pointing (I have no idea about this), then it might still be worth keeping this focal plane coordinate system around and using it for this purpose.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://github.com/rmjarvis/Piff/pull/24/files/02a1fef3b765a4d95d314f43082b4a18f76706d2#r69611430>, or mute the thread<https://github.com/notifications/unsubscribe/AEEa87SKNQpvyp1QhAHJ_3QpNVLqpJ0zks5qSp8XgaJpZM4I6RGE>.

_________________________________________________________________________
 Prof. Aaron Roodman
 Chair Dept. of Particle Physics & Astrophysics
 SLAC National Accelerator Laboratory
 Stanford University


 SLAC National Accelerator Laboratory        E-mail: roodman@slac.stanford.edu<mailto:roodman@slac.stanford.edu>
 2575 Sand Hill Rd.                                           Phone:  650-926-2705
 MS 29                                                                  Fax:  650-926-2657
 Menlo Park, CA 94025                                    URL: http://www.slac.stanford.edu/~roodman
_________________________________________________________________________